src/prng/hc128.rs

```diff
@@ -0,0 +1,543 @@
+            self.t[cc+512+9]  = self.step_q(cc+9,  cc+10, cc+6,  ee+15, ee+13);
+            self.t[cc+512+10] = self.step_q(cc+10, cc+11, cc+7,  cc+0,  ee+14);
+            self.t[cc+512+11] = self.step_q(cc+11, cc+12, cc+8,  cc+1,  ee+15);
+            self.t[cc+512+12] = self.step_q(cc+12, cc+13, cc+9,  cc+2,  cc+0);
+            self.t[cc+512+13] = self.step_q(cc+13, cc+14, cc+10, cc+3,  cc+1);
+            self.t[cc+512+14] = self.step_q(cc+14, cc+15, cc+11, cc+4,  cc+2);
+            self.t[cc+512+15] = self.step_q(cc+15, dd+0,  cc+12, cc+5,  cc+3);
+        }
+        self.counter1024 += 16;
+    }
+}
+
+impl Rng for Hc128Rng {
+    #[inline]
+    fn next_u32(&mut self) -> u32 {
+        if self.index >= 16 {
+            self.state.update(&mut self.results);
+            self.index = 0;
+        }
+
+        let value = self.results[self.index];
+        self.index += 1;
+        value
+    }
+
+    #[inline]
+    fn next_u64(&mut self) -> u64 {
+        let index = self.index;
+        if index < 15 {
+            self.index += 2;
+            // Read an u64 from the current index
+            if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
+                unsafe { *(&self.results[index] as *const u32 as *const u64) }
+            } else {
+                let x = self.results[index] as u64;
+                let y = self.results[index + 1] as u64;
+                (y << 32) | x
+            }
+        } else if index >= 16 {
+            self.state.update(&mut self.results);
+            self.index = 2;
+            let x = self.results[0] as u64;
+            let y = self.results[1] as u64;
+            (y << 32) | x
+        } else {
+            let x = self.results[15] as u64;
+            self.state.update(&mut self.results);
+            self.index = 1;
+            let y = self.results[0] as u64;
+            (y << 32) | x
+        }
+    }
+
+    // As an optimization we try to write directly into the output buffer.
+    // This is only enabled for platforms where unaligned writes are known to
+    // be safe and fast.
+    // This improves performance by about 12%.
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        let mut filled = 0;
+
+        // Continue filling from the current set of results
+        if self.index < 16 {
+            let (consumed_u32, filled_u8) =
+                impls::fill_via_u32_chunks(&mut self.results[self.index..],
+                                           dest);
+
+            self.index += consumed_u32;
+            filled += filled_u8;
+        }
+
+        let len_remainder = (dest.len() - filled) % 16;
+        let len_direct = dest.len() - len_remainder;
+
+        while filled < len_direct {
+            let dest_u32: &mut [u32] = unsafe {
+                slice::from_raw_parts_mut(
+                    dest[filled..].as_mut_ptr() as *mut u8 as *mut u32,
+                                      16)
```

Comment:
I've used the `array_ref!` macro a few times myself. Maybe we really should copy it into `rand`? Not important.

Ah, this is where unaligned copies happen.

Reply:
I wrote this before we had it. I don't think about it enough yet to be honest, maybe just have to see it a few more times in action. Should I modify this?

Reply:
No, `array_ref` is not in `rand` yet. Maybe we should add it; not sure, and not sure if it should be exported. It would probably be more useful in `rand-core` if we do export the macro.